pkg/syncer/metadatasyncer.go

```diff
@@ -551,6 +573,198 @@ func InitMetadataSyncer(ctx context.Context, clusterFlavor cnstypes.CnsClusterFl
+	var nodeTopoObj csinodetopologyv1alpha1.CSINodeTopology
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.(*unstructured.Unstructured).Object, &nodeTopoObj)
+	if err != nil {
+		log.Errorf("topoCRAdded: failed to cast object %+v to %s. Error: %v", obj,
+			csinodetopology.CRDSingular, err)
+		return
+	}
+	// Check if Status is set to Success.
+	if nodeTopoObj.Status.Status != csinodetopologyv1alpha1.CSINodeTopologySuccess {
+		log.Infof("topoCRAdded: CSINodeTopology instance %q not yet ready. Status: %q",
+			nodeTopoObj.Name, nodeTopoObj.Status.Status)
+		return
+	}
+	addLabelsToTopologyVCMap(ctx, nodeTopoObj)
+}
+
+// topoCRUpdated checks if the CSINodeTopology instance Status is set to Success
+// and populates the metadataSyncer.topologyVCMap with appropriate values.
+func topoCRUpdated(oldObj interface{}, newObj interface{}) {
+	ctx, log := logger.GetNewContextWithLogger()
+	// Verify both objects received.
+	var (
+		oldNodeTopoObj csinodetopologyv1alpha1.CSINodeTopology
+		newNodeTopoObj csinodetopologyv1alpha1.CSINodeTopology
+	)
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(
+		newObj.(*unstructured.Unstructured).Object, &newNodeTopoObj)
+	if err != nil {
+		log.Errorf("topoCRUpdated: failed to cast new object %+v to %s. Error: %+v", newObj,
+			csinodetopology.CRDSingular, err)
+		return
+	}
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(
+		oldObj.(*unstructured.Unstructured).Object, &oldNodeTopoObj)
+	if err != nil {
+		log.Errorf("topoCRUpdated: failed to cast old object %+v to %s. Error: %+v", oldObj,
+			csinodetopology.CRDSingular, err)
+		return
+	}
+	// If status has been updated to Success, populate the metadataSyncer.topologyVCMap.
+	if newNodeTopoObj.Status.Status == csinodetopologyv1alpha1.CSINodeTopologySuccess {
+		if oldNodeTopoObj.Status.Status != csinodetopologyv1alpha1.CSINodeTopologySuccess {
+			addLabelsToTopologyVCMap(ctx, newNodeTopoObj)
+		} else {
+			// Both old and updated CSINodeTopology instances were successful.
+			// Check if there is any change in the topology labels.
+			oldTopoLabelsMap := make(map[string]string)
+			for _, label := range oldNodeTopoObj.Status.TopologyLabels {
+				oldTopoLabelsMap[label.Key] = label.Value
+			}
+			newTopoLabelsMap := make(map[string]string)
+			for _, label := range newNodeTopoObj.Status.TopologyLabels {
+				newTopoLabelsMap[label.Key] = label.Value
+			}
+			// Check if there are updates to the topology labels in the Status.
+			if reflect.DeepEqual(oldTopoLabelsMap, newTopoLabelsMap) {
+				log.Debugf("topoCRUpdated: No change in %s CR topology labels. Ignoring the event",
+					csinodetopology.CRDSingular)
+				return
+			}
+			log.Warnf("topoCRUpdated: %q instance with name %q has been updated after the Status was set to "+
+				"Success. Old object - %+v. New object - %+v", csinodetopology.CRDSingular, oldNodeTopoObj.Name,
+				oldNodeTopoObj, newNodeTopoObj)
+			// TODO: Update metadataSyncer.topologyVCMap?
```

Comment:
We do not support updating Node Topology.

Reply:
We discussed that we will update the cache to keep the correctness WRT K8s knowledge, changed the logic and removed TODO.